examples/vortex-mpi.py

```diff
@@ -149,54 +150,57 @@ def main(ctx_factory=cl.create_some_context, use_profiling=False, use_logmgr=Fal
                                      nelements=local_nelements,
-                                     global_nelements=global_nelements,
-                                     dt=current_dt, t_final=t_final, nstatus=nstatus,
-                                     nviz=nviz, cfl=current_cfl,
-                                     constant_cfl=constant_cfl, initname=initname,
-                                     eosname=eosname, casename=casename)
+                                     global_nelements=global_nelements)
     if rank == 0:
         logger.info(init_message)
 
-    get_timestep = partial(inviscid_sim_timestep, discr=discr, t=current_t,
-                           dt=current_dt, cfl=current_cfl, eos=eos,
-                           t_final=t_final, constant_cfl=constant_cfl)
+    def get_timestep(t, state):
+        dt = get_inviscid_timestep(discr=discr, q=state, cfl=current_cfl,
+            eos=eos) if constant_cfl else current_dt
+        return min(dt, t_final - t)
 
-    def my_rhs(t, state):
+    def rhs(t, state):
         return inviscid_operator(discr, q=state, t=t,
                                  boundaries=boundaries, eos=eos)
 
-    def my_checkpoint(step, t, dt, state):
-        return sim_checkpoint(discr, visualizer, eos, q=state,
-                              exact_soln=initializer, vizname=casename, step=step,
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              exittol=exittol, constant_cfl=constant_cfl, comm=comm,
-                              vis_timer=vis_timer)
-
-    try:
-        (current_step, current_t, current_state) = \
-            advance_state(rhs=my_rhs, timestepper=timestepper,
-                          checkpoint=my_checkpoint,
-                          get_timestep=get_timestep, state=current_state,
-                          t=current_t, t_final=t_final, logmgr=logmgr, eos=eos,
-                          dim=dim)
-    except ExactSolutionMismatch as ex:
-        current_step = ex.step
-        current_t = ex.t
-        current_state = ex.state
-
-    #    if current_t != checkpoint_t:
+    def write_vis(step, t, state):
+        cv = split_conserved(dim, state)
+        exact_cv = split_conserved(dim, initializer(nodes, t=t))
+        io_fields = [
+            ("cv", cv),
+            ("dv", eos.dependent_vars(cv)),
+            ("exact_cv", exact_cv),
+        ]
+        return write_visualization_file(visualizer, fields=io_fields,
+                    basename=casename, step=step, t=t, comm=comm, timer=vis_timer)
+
+    def checkpoint(step, t, dt, state):
+        done = t >= t_final
+        exact_state = initializer(nodes, t=t)
+        if comm_any(comm, discr.norm(state - exact_state, np.inf) > exittol):
+            write_vis(step, t, state)
+            raise RuntimeError(f"Exact solution mismatch at step {step}.")
+        if check_step(step, nviz) or done:
+            write_vis(step, t, state)
+        return done
```

Comment:
I do like how this is shaping up - but I am still concerned that this leaves a lot to the user to get right in every driver.  I don't think I found a default or checkpoint helper. The same seems to be true for the `write_viz` functionality above.  

Can we provide some built-in defaults for these so that the user just needs to "overload" them if they need to customize?